README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
readnet
=======

The goal of readnet is to provide tools for preprocessing network data (a companion to igraph, statnet, netdiffuseR, etc.)

Some of the features:

-   Automatically handle coding for nodes, time and groups.

-   If either `group.id` or `time` are specified, the function makes recursive calls, returning nested edgelists with the following structure:

        object
        --- group1:
          --- edgelist in time1
          --- ...
          --- edgelist in time n
        ...
        --- group n:
          --- edgelist in time1
          --- ...
          --- edgelist in time n

-   Includes functions to coerce the resulting edgelists into igraph objects and viceversa.

This project is on development and not published on CRAN yet.

Installation
------------

You can get `readnet` using devtools:

``` r
devtools::install_github("USCCANA/readnet")
```

Example
-------

Survey to Edgelist
------------------

This is a basic example which shows you how to solve a common problem. Suppose that you have a dataset that looks something like this:

``` r
dat
#>   ego alter1 alter2 time
#> 1   a      b   <NA>    0
#> 2   b   <NA>   <NA>    0
#> 3   c   <NA>   <NA>    0
#> 4   a      b      c    1
#> 5   b      a   <NA>    1
#> 6   c   <NA>   <NA>    1
```

This is, a survey with network nominations, and in this case, including a time variable. We can use the function `survey_to_edgelist` to process this data

``` r
library(readnet)
(ans <- survey_to_edgelist(dat, "ego", c("alter1", "alter2"), time = "time"))
#> $`0`
#> $edgelist
#>      ego alter
#> [1,]   1     2
#> 
#> $labels
#> [1] "a" "b" "c"
#> 
#> $data
#>   ego alter1 alter2 time
#> 1   a      b   <NA>    0
#> 2   b   <NA>   <NA>    0
#> 3   c   <NA>   <NA>    0
#> 
#> $graph.attrs
#> NULL
#> 
#> attr(,"class")
#> [1] "rn_edgelist"
#> 
#> $`1`
#> $edgelist
#>      ego alter
#> [1,]   1     2
#> [2,]   2     1
#> [3,]   1     3
#> 
#> $labels
#> [1] "a" "b" "c"
#> 
#> $data
#>   ego alter1 alter2 time
#> 4   a      b      c    1
#> 5   b      a   <NA>    1
#> 6   c   <NA>   <NA>    1
#> 
#> $graph.attrs
#> NULL
#> 
#> attr(,"class")
#> [1] "rn_edgelist"
```

The previous returned a list with elements of class `rn_edgelist`. Each element contains an edgelist, a vector of labels, and a dataframe with data. We can use the function `as_igraph` to turn this into an igraph object

``` r
as_igraph(ans)
#> $`0`
#> IGRAPH 8103e75 DN-- 3 1 -- 
#> + attr: name (v/c), ego (v/c), alter1 (v/c), alter2 (v/c), time
#> | (v/c)
#> + edge from 8103e75 (vertex names):
#> [1] a->b
#> 
#> $`1`
#> IGRAPH d3b3db4 DN-- 3 3 -- 
#> + attr: name (v/c), ego (v/c), alter1 (v/c), alter2 (v/c), time
#> | (v/c)
#> + edges from d3b3db4 (vertex names):
#> [1] a->b b->a a->c
```

and as a `rn_edgelist` back

``` r
as_rn_edgelist(as_igraph(ans))
#> $`0`
#> $edgelist
#>      ego alter
#> [1,]   1     2
#> 
#> $labels
#> [1] "a" "b" "c"
#> 
#> $data
#>   name ego alter1 alter2 time
#> a    a   a      b   <NA>    0
#> b    b   b   <NA>   <NA>    0
#> c    c   c   <NA>   <NA>    0
#> 
#> $graph.attrs
#> named list()
#> 
#> attr(,"class")
#> [1] "rn_edgelist"
#> 
#> $`1`
#> $edgelist
#>      ego alter
#> [1,]   1     2
#> [2,]   2     1
#> [3,]   1     3
#> 
#> $labels
#> [1] "a" "b" "c"
#> 
#> $data
#>   name ego alter1 alter2 time
#> a    a   a      b      c    1
#> b    b   b      a   <NA>    1
#> c    c   c   <NA>   <NA>    1
#> 
#> $graph.attrs
#> named list()
#> 
#> attr(,"class")
#> [1] "rn_edgelist"
```

Setting up data for Siena (or alike :))
---------------------------------------

Here is an example that I built for [Prof. Kayla de la Haye](http://profiles.sc-ctsi.org/kayla.delahaye/). The idea here is that we have an edgelist with (finite) timestamps, a dataset with node attributes, and a list of ids that we would like to include in the data. The final output looks like lists of adjacency matrices of *n* × *n* (so we coded the edgelist as 1 through `n`) and `data.frame` of the attributes matching the positions of nodes in the edgelist (you may be able to do something similar with igraph, but the key difference here is that this function allows you to include time and returns the attribute dataset sorted accordignly):

``` r
library(readnet)
library(dplyr)

edgelist <- readxl::read_excel("playground/kayla/GROW_Edgelist clean for George.xlsx")
dat      <- readr::read_tsv("playground/kayla/GROW Attribute Data for George.csv")
ids2work <- readxl::read_excel("playground/kayla/GROW_Edgelist clean for George.xlsx", "NetMembersN400")


# Catched an error:
# Error: Missing ids in `data`. There are 610 observations in `data` and 613
# identified ids. The observations that are missing in `data` are: '812398',
# '812660', '831086'.

# The same error shows up when I run it filtering the data:
#
#  Error: Missing ids in `data`. There are 394 observations in `data` and 397
# identified ids. The observations that are missing in `data` are: '812398',
# '812660', '831086'. So
#
# So I'll just drop it!

dids <- unique(dat$study_id)
edgelist <- filter(edgelist, (Sender %in% dids) & (Receiver %in% dids))

ans <- edgelist_to_adjmat_w_attributes(
  edgelist         = edgelist,
  data             = dat,
  data.idvar       = "study_id",
  edgelist.timevar = "wave",
  ids.to.keep      = ids2work$study_id
  )
```


ChangeLog
2018-01-12  George G. Vega Yon <g.vegayon@gmail.com>

	* : Started changelog
	* : Added example using Kayla's data (without the data of course).


tests/testthat.R
library(testthat)
library(readnet)

test_check("readnet")


tests/testthat/test-readnet.r
context("readnet.r")

test_that("Basic test", {

  dat0 <- data.frame(
    ego    = c("a", "b", "c"),
    alter1 = c("b", NA, NA),
    alter2 = NA,
    time   = 0
  )

  dat1 <- data.frame(
    ego    = c("a", "b", "c"),
    alter1 = c("b", "a", NA),
    alter2 = c("c", NA, NA),
    time   = 1
    )


  ans0 <- structure(list(
    `0` = structure(
      list(
        edgelist = structure(1:2, .Dim = 1:2, .Dimnames = list(NULL, c("ego", "alter"))),
        labels = c("a", "b", "c"),
        data = structure(
          list(
            ego = c("a", "b", "c"),
            alter1 = c("b", NA, NA),
            alter2 = c(NA_character_,
                       NA_character_, NA_character_),
            time = c("0", "0", "0")
          ),
          .Names = c("ego",
                     "alter1", "alter2", "time"),
          row.names = c(NA, 3L),
          class = "data.frame"
        ),
        graph.attrs=NULL
      ),
      class = "rn_edgelist"
    ),
    `1` = structure(
      list(
        edgelist = structure(
          c(1L, 2L, 1L, 2L, 1L, 3L),
          .Dim = c(3L,
                   2L),
          .Dimnames = list(NULL, c("ego", "alter"))
        ),
        labels = c("a",
                   "b", "c"),
        data = structure(
          list(
            ego = c("a", "b", "c"),
            alter1 = c("b", "a", NA),
            alter2 = c("c", NA, NA),
            time = c("1",
                     "1", "1")
          ),
          .Names = c("ego", "alter1", "alter2", "time"),
          row.names = 4:6,
          class = "data.frame"
        ),
        graph.attrs = NULL
      ),
      class = "rn_edgelist"
    )
  ), .Names = c("0", "1"))

  ans <- survey_to_edgelist(rbind(dat0, dat1), "ego", c("alter1", "alter2"), time = "time")
  expect_equal(ans, ans0)
})

test_that("igraph", {

  set.seed(12)
  ans0 <- igraph::barabasi.game(10)
  igraph::vertex_attr(ans0, "name") <- as.character(1L:10L)
  igraph::V(ans0)$x <- runif(10)

  ans1 <- as_igraph(as_rn_edgelist(ans0))

  expect_equal(igraph::as_adj(ans0), igraph::as_adj(ans1))
  expect_equal(igraph::graph_attr(ans0), igraph::graph_attr(ans1))
  expect_equal(igraph::edge_attr(ans0), igraph::edge_attr(ans1))
  expect_equal(igraph::vertex_attr(ans0), igraph::vertex_attr(ans1))

})


R/eval.r
library(magrittr)

new_graph <- function(.data) {
  env <- new.env()
  env[["DF"]] <- .data
  env
}

getval <- function(x) {
  as.character(substitute(x, parent.frame()))
}

set_variable <- function(.data, var, name) {

  # Adds the variable to the list... if it exists
  .data[[name]] <- c(
    .data[[name]],
    var
  )

  # Checks whether the variable exists
  if (!length(.data$DF[[.data[[name]]]]))
    stop("The variable `", .data[[name]], "` does not exists.", call. = FALSE)

  .data
}

set_group <- function(.data, var) {
  set_variable(.data, getval(var), "group")
}

set_time <- function(.data, var) {
  set_variable(.data, getval(var), "time")
}

get_edgelist <- function(.data, ego, ...) {

  set_variable(.data, ego, "ego")

  for (alter in list(...)) {
    set_variable(.data, getval(var), "alter")
  }

  .data
}

set_edgelist <- function(.data, edgelist) {

}

# dat0 <- data.frame(
#   ego    = c("a", "b", "c"),
#   alter1 = c("b", NA, NA),
#   alter2 = NA,
#   time   = 0
# )
#
# dat1 <- data.frame(
#   ego    = c("a", "b", "c"),
#   alter1 = c("b", "a", NA),
#   alter2 = c("c", NA, NA),
#   time   = 1
# )
#
# dat <- rbind(dat0, dat1)
#
# ans <- dat %>%
#   new_graph %>%
#   set_group(time) %>%
#   get_edgelist(ego, alter1, alter2)


R/misc.r
#' This function takes a two-column matrix, generates factors and encodeds it
#' into an integer matrix with an extra attribute: `labels`
#' @noRd
encode_mat <- function(x) {

  x      <- as.factor(c(x[, 1L], x[, 2L]))
  labels <- levels(x)

  x <- matrix(as.integer(x), ncol=2, byrow = FALSE,
              dimnames = list(NULL, c("ego", "alter")))
  attr(x, "labels") <- labels

  x

}

#' This function takes a set of vectors and creates a unique id using all
#' by pasting it.
#' @noRd
makegroups <- function(dat, group.id, sep) {
  apply(dat[, group.id, drop=FALSE], 1, paste0, collapse=sep)
}

#' This function coerces a set of variables in a data frame into character
#' so that when that is turned into a factor, information is not loss.
#' @noRd
set_as_char <- function(dat, vars) {

  dat <- as.character(match.call()$dat)
  env <- parent.frame()

  for (v in vars)
    if (!is.character(env[[dat]][[v]]))
      env[[dat]][[v]] <- as.character(env[[dat]][[v]])
}


R/readnet.r
#' Function to coerce an object to class `edgelist`
#' @noRd
new_rn_edgelist <- function(
  edgelist,
  labels,
  data,
  graph.attrs = NULL,
  ego.id = NULL,
  checks = c("edgelist", "labels", "data")
) {

  # Automatic coercion
  edgelist   <- unname(as.matrix(edgelist))
  edgelist[] <- as.integer(edgelist[])
  colnames(edgelist) <- c("ego", "alter")

  labels     <- unname(as.character(unlist(labels, recursive = TRUE)))

  data       <- as.data.frame(data)

  n <- length(labels)

  if ("labels" %in% checks)
    if (any(is.na(labels)))
      stop("Some `labels` are missing.", call. = FALSE)

  if ("data" %in% checks)
    if (nrow(data) != n)
      stop("The number of observations in `data`, ", nrow(data), ", does not",
           "  coincide with the number of labels extracted from the edgelist, ",
           n, ". Yo probably have duplicated observations in `data`.", call. = FALSE)


  if ("edgelist" %in% checks) {

    # Complete cases
    test <- which(!complete.cases(edgelist))
    if (length(test))
      stop("Some elements of the `edgelist` are missing.", call. = FALSE)

    # Pointing to the labels
    test <- range(edgelist)
    if (test[1] < 0 | test[2] > n)
      stop("Some elements of the `edgelist` have id values out of range (1 or n).",
           call. = FALSE)

  }

  # If to be sorted
  if (!length(ego.id))
    data <- data[match(labels, data[[ego.id]]), , drop=FALSE]


  # Creating the object
  structure(
    list(
      edgelist    = edgelist,
      labels      = labels,
      data        = data,
      graph.attrs = graph.attrs
    ),
    class = "rn_edgelist"
  )

}

#' Creates an edgelist (or a list of edgelist) from survey data
#'
#' The function grabs ego ids and its nominations and creates an edgelist
#' in which the labels (ids) are recoded such that ids go from 1 to n.
#'
#' @param data A `data.frame` with the data to be read in.
#' @param ego.id,alter.id Character scalars. Names of the variables in `dat` that
#' correspond to the ids of ego and alter. It does not need to be numeric
#' (see details).
#' @param group.id Character vector.
#' @param time Character vector.
#' @param group.sep Character scalar.
#' @param time.sep Character scalar.
#'
#' @details
#' By default, `ego.id` and `alter.id` are transform to factor variables, so the
#' user doesn't need to worry about giving individuals in the survey a special
#' coding
#'
#' @return If no `group.id` or `time` are provided, an integer matrix with
#' two named columns (_ego_ and _alter_), and an attribute "`labels`". Otherwise
#' will return a nested list in which the first level is given by `group.id`,
#' and the second level by `time` (if provided).
#'
#' @examples
#'
#' 1+1
#' @export
#' @aliases rn_edgelist
survey_to_edgelist <- function(
  data,
  ego.id,
  alter.id,
  group.id  = NULL,
  time      = NULL,
  group.sep = "-",
  time.sep  = "-"
  ) {

  # Setting nasty option that I don't want to use!
  old_strasf <- getOption("stringsAsFactors")
  on.exit(options(stringsAsFactors = old_strasf))
  options(stringsAsFactors = FALSE)

  # Making sure everything is character
  set_as_char(data, c(ego.id, alter.id, group.id, time))

  # Checking if group was specified, then call it recursively
  if (length(group.id)) {

    # Group tags
    group.id <- makegroups(data, group.id, group.sep)

    # Calling the function recursively
    return(
      lapply(
        split.data.frame(data, group.id),
        survey_to_edgelist,
        ego.id = ego.id, alter.id = alter.id, group.id = NULL, time = time)
      )
  }

  # Creates some containers
  n          <- nrow(data)
  k          <- length(alter.id)
  ans        <- vector("list", k)
  names(ans) <- alter.id

  # Pastes the data
  for (id in alter.id) {
    ans[[id]] <- cbind(data[[ego.id]], data[[id]])
  }

  ans <- encode_mat(do.call(rbind, ans))
  labels <- attr(ans, "labels")

  # If time was provided
  if (length(time)) {
    time <- makegroups(data, time, time.sep)
    ans  <- split.data.frame(ans, time)
    data  <- split.data.frame(data, time)

    # Mapping the function
    ans  <- Map(function(e, d) {

      new_rn_edgelist(
        edgelist = e[stats::complete.cases(e),,drop=FALSE],
        labels   = labels,
        data     = d,
        ego.id   = ego.id,
        checks   = "data"
      )

    }, e = ans, d = data)

  } else {

    ans <- new_rn_edgelist(
      edgelist = ans[stats::complete.cases(ans),,drop=FALSE],
      labels   = labels,
      data     = data,
      ego.id   = ego.id,
      checks   = "data"
    )

  }

  # Returning including the data
  ans
}

#' Coerce objects `rn_edgelist` to `igraph`
#' @param x An object of class `rn_edgelist` or a list of that.
#' @param ... Further arguments passed to [`graph_from_data_frame`][igraph::graph_from_data_frame()]
#' from the \pkg{igraph} R package.
#' @export
as_igraph <- function(x, ...)
  UseMethod("as_igraph")

#' @export
as_igraph.list <- function(x, ...) {
  lapply(x, as_igraph, ...)
}

#' @export
as_igraph.rn_edgelist <- function(x, ...) {

  # Creating the vertex matrix
  vertices <- cbind(
    data.frame(name = x$labels),
    x$data
  )

  # Relabeling the edgelist...
  edgelist   <- x$edgelist
  edgelist[] <- x$labels[edgelist[]]

  # Putting all together
  ans <- igraph::graph_from_data_frame(edgelist, vertices = vertices, ...)

  for (a in names(x$graph.attr))
    igraph::graph_attr(ans, a) <- x$graph.attr[[a]]

  ans

}

#' Coerce different graph objects to `rn_edgelist`
#' @param x An object to be coerced to [rn_edgelist]
#' @export
as_rn_edgelist <- function(x) UseMethod("as_rn_edgelist")

#' @export
as_rn_edgelist.list <- function(x)
  lapply(x, as_rn_edgelist)

#' @export
as_rn_edgelist.igraph <- function(x) {

  labels <- igraph::vertex_attr(x, "name")
  if (!length(labels))
    labels <- as.character(1L:igraph::vcount(x))

  new_rn_edgelist(
    edgelist = igraph::as_edgelist(x, names = FALSE),
    labels   = labels,
    data     = igraph::as_data_frame(x, "vertices"),
    graph.attrs = igraph::graph_attr(x),
    ego.id   = "name"
  )


}


R/edgelist.r
#' Basic checks of the data passed.
#' @param data A data.frame
#' @noRd
check_data <- function(data, idvar, timevar = NULL) {

  # Is it a data.frame
  if (!is.data.frame(data))
    stop("`data` must be an object of class `data.frame`.", call. = FALSE)

  # Is it unique? --------------------------------------------------------------
  if (length(timevar)) {

    if (!length(data[[timevar]]))
      stop("The variable '", timevar, "' does not exists in `data`.", call. = FALSE)

    split(data[[timevar]], check_data, idvar = idvar)
  }

  if (!length(data[[idvar]]))
    stop("The variable '", idvar, "' does not exists in `data`.", call. = FALSE)

  if (any(duplicated(data[[idvar]])))
    stop("The panel has duplicated data.", call. = FALSE)

}

edgelist_to_adjmat <- function(
  edgelist,
  data     = NULL,
  ego.id   = NULL,
  group.id = NULL,
  timevar  = NULL
) {

  # Setting nasty option that I don't want to use!
  old_strasf <- getOption("stringsAsFactors")
  on.exit(options(stringsAsFactors = old_strasf))
  options(stringsAsFactors = FALSE)

  # Checks
  if (!did_i_call_myself("edgelist_to_adjmat")) {
    check_data(data, idvar, timevar)

    # Coercing into character
    set_as_char(data, c(idvar, timevar, ego.id, group.id, timevar))

    nodes <- data[[idvar]]
    nodes <- data.frame(id = 1L:length(node), label = nodes)

    # Are all the fellas of `edgelist` in `data`?
    test <- setdiff(as.vector(edgelist), nodes[["label"]])
    if (length(test))
      stop("The following nodes do not show in `data`:\n",
           paste(unique(test), collapse = ", "), ".", call. = FALSE)
  } else {
    env <- parent.frame()
    nodes <- env[["nodes"]]
  }

  if (length(timevar)) {
    Map(function(e, d) {
      edgelist_to_adjmat(e, d, ego.id = egp.id, group.id = group.id)
    },
    e = split.data.frame(data, data[[timevar]]),
    d = split.data.frame(data, data[[timevar]])
      ,
      edgelist_to_adjmat,

    )
  }

  # Recoding edgelist
  edgelist[, 1] <- nodes[match(edgelist[, 1], nodes[, 2]), 1]
  edgelist[, 2] <- nodes[match(edgelist[, 2], nodes[, 2]), 1]

  # Creating the adjmat
  adjmat <- matrix()


}

#' Returns `TRUE` if the function `fn` was called by itself (recursion)
#' @noRd
did_i_call_myself <- function(fn) {
  isrecur <- as.character(sys.call(-1)[[1]]) == fn
  if (length(isrecur) && isrecur)
    return(TRUE)
  else (FALSE)
}




R/encode_together.r

#' Creates an adjacency matrix from an edgelist with vertex attributes
#'
#' The only difference with [igraph::graph_from_data_frame] is that
#' this returns an 0/1 adjacency matrix and sorts whatever attributes were
#' passed via `data` accordingly to the network coding.
#'
#' @param edgelist A data.frame. Only the first two columns will be used.
#' @param data A data.frame with vertex attributes.
#' @param data.idvar Character scalar. Name of the variable that holds the
#' vertex id (label). By default it uses the first column.
#'
#' @return A list with the following
#' \item{adjmat}{A matrix of \eqn{n\times n}.}
#' \item{data}{A sorted version of `data` according to the network coding.}
#'
#' @examples
#' # Simple example
#' net <- c("a", "b", "a", "c", "b", "c", "c", "b", "c", "a")
#' net <- matrix(net, byrow = TRUE, ncol=2)
#'
#' dat <- data.frame(name = c("d", "a", "c", "b"), age = c(4, 1, 3, 2))
#' ans0 <- edgelist_to_adjmat_w_attributes(net, dat)
#' @name edgelist_to_adjmat_w_attributes
NULL

#' Basically, this function replicates a dataframe so it
#' @noRd
time_check_and_fill <- function() {

  # We will be working with the data of the parent function
  env <- parent.frame()

  # Finding time ---------------------------------------------------------------
  times.e <- if (length(env$edgelist.timevar)) {
    unique(env[["edgelist"]][[env$edgelist.timevar]])
  } else {
    NULL
  }

  times.d <- if (length(env$data.timevar)) {
    unique(env[["data"]][[env$data.timevar]])
  } else {
    NULL
  }
  times <- sort(unique(c(times.e, times.d)))

  # If defined All times should be present -------------------------------------

  # In the edgelist
  test <- times[which(!(times %in% times.e))]
  if (length(times.e) && length(test))
    stop("Some time periods in `data` are not present in `edgelist`: ",
         paste(test, collapse=", "), ".")

  # In the data
  test <- times[which(!(times %in% times.d))]
  if (length(times.d) && length(test))
    stop("Some time periods in `edgelist` are not present in `data`: ",
         paste(test, collapse=", "), ".")

  # If no time is specified for one, then fill the gaps ------------------------

  # For the edgelist
  if (!length(env[["edgelist.timevar"]])) {
    env[["edgelist"]] <- do.call(
      rbind,
      lapply(times, function(i) cbind(env[["edgelist"]], time = i))
      )

    env[["edgelist.timevar"]] <- "time"
  }

  # For the data
  if (!length(env[["data.timevar"]])) {
    env[["data"]] <- do.call(
      rbind,
      lapply(times, function(i) cbind(env[["data"]], time = i))
    )

    env[["data.timevar"]] <- "time"
  }

}

#' @rdname edgelist_to_adjmat_w_attributes
#' @export
edgelist_to_adjmat_w_attributes <- function(
  edgelist,
  data,
  data.idvar       = colnames(data)[1],
  ids.to.keep      = NULL,
  data.timevar     = NULL,
  edgelist.timevar = NULL
) {

  # Here goes checks
  # 1. Both are data.frames
  # 2. edgelist has two columns
  # 3. attributes must be unique!

  edgelist <- as.data.frame(edgelist)
  data     <- as.data.frame(data)

  # Filtering ids
  if (length(ids.to.keep)) {

    edgelist <- edgelist[
      (edgelist[[1]] %in% ids.to.keep) & (edgelist[[2]] %in% ids.to.keep),
      ]

    data <- data[data[[data.idvar]] %in% ids.to.keep, ]

  }

  # Apply recursively if timevar are specified
  if (length(edgelist.timevar) | length(data.timevar)) {

    # Finding time
    times <- NULL
    time_check_and_fill()

    ans <- Map(
      function(e, d) {
        edgelist_to_adjmat_w_attributes(
          edgelist    = e,
          data        = d,
          data.idvar  = data.idvar
          )
        },
      e = split.data.frame(edgelist, edgelist[[edgelist.timevar]]),
      d = split.data.frame(data, data[[data.timevar]])
      )

    names(ans) <- times

    return(ans)


  }

  # Creating dictionary
  charedgelist <- as.vector(unlist(edgelist[,1:2]))
  dict <- sort(unique(c(
    charedgelist,
    as.character(data[[data.idvar]])
    )))

  n <- length(dict)

  # There should not be any duplicated data
  if (any(duplicated(data[[data.idvar]])))
    stop("`data` has duplicated observations.", call. = FALSE)

  # This all should match
  if (n != nrow(data))
    stop("Missing ids in `data`. There are ", nrow(data),
         " observations in `data` and ", n, " identified ids. ",
         "The observations that are missing in `data` are: '",
         paste(setdiff(dict, data[[data.idvar]]), collapse="', '"), "'.",
         call. = FALSE)

  # Encoding edgelist
  new_edgelist <- matrix(0L, nrow=nrow(edgelist), ncol=2)
  new_edgelist[] <- (1:n)[match(charedgelist, dict)]

  # Encoding data (sorting)
  data <- data[order(match(data[[data.idvar]], dict)),]

  # Returning

  adjmat <- matrix(0, nrow=n, ncol=n, dimnames = list(dict, dict))
  adjmat[new_edgelist] <- 1L
  list(
    adjmat = adjmat,
    data   = data
  )

}


playground/kayla/grow.r
rm(list = ls())

library(readnet)
library(dplyr)

edgelist <- readxl::read_excel("playground/kayla/GROW_Edgelist clean for George.xlsx")
dat      <- readr::read_tsv("playground/kayla/GROW Attribute Data for George.csv")
ids2work <- readxl::read_excel("playground/kayla/GROW_Edgelist clean for George.xlsx", "NetMembersN400")


# Catched an error:
# Error: Missing ids in `data`. There are 610 observations in `data` and 613
# identified ids. The observations that are missing in `data` are: '812398',
# '812660', '831086'.

# The same error shows up when I run it filtering the data:
#
#  Error: Missing ids in `data`. There are 394 observations in `data` and 397
# identified ids. The observations that are missing in `data` are: '812398',
# '812660', '831086'. So
#
# So I'll just drop it!

dids <- unique(dat$study_id)
edgelist <- filter(edgelist, (Sender %in% dids) & (Receiver %in% dids))

ans <- edgelist_to_adjmat_w_attributes(
  edgelist         = edgelist,
  data             = dat,
  data.idvar       = "study_id",
  edgelist.timevar = "wave",
  ids.to.keep      = ids2work$study_id
  )
